File: mpp/lab4/server/src/main/java/repository/ParticipantRepository.java
package repository;

import domain.Participant;

import java.sql.*;

public class ParticipantRepository extends DatabaseRepository
        implements IParticipantRepository {
    public ParticipantRepository(Connection connection) throws RepositoryError {
        super(connection);
    }

    @Override
    public void createTable() throws RepositoryError {
        if (isTableCreated("Participants")) {
            return;
        }

        String command =
                "create table Participants(" +
                    "ParticipantId integer primary key autoincrement, " +
                    "ParticipantName nvarchar(255) not null unique" +
                ");";

        try {
            Statement statement = connection.createStatement();
            statement.execute(command);
            logger.info("Created table Participants");
            logger.info(command);
        } catch (SQLException e) {
            logger.error("Failed to create table Participants");
            logger.error(e);
            throw new RepositoryError(e);
        }
    }

    @Override
    public void add(Participant participant) throws RepositoryError {
        String command =
                "insert into Participants (" +
                    "ParticipantName" +
                ") values (" +
                    "?" +
                ");";

        try {
            PreparedStatement statement = connection.prepareStatement(command, Statement.RETURN_GENERATED_KEYS);
            statement.setString(1, participant.getName());
            statement.executeUpdate();

            ResultSet results = statement.getGeneratedKeys();
            if (results.next()) {
                participant.setId(results.getInt(1));
            } else {
                throw new RepositoryError("Failed to retrieve generated keys");
            }

            logger.info("Added participant");
            logger.info(command);
        } catch (SQLException e) {
            logger.error("Failed to add participant");
            logger.error(command);
            logger.error(e);
            throw new RepositoryError(e);
        }
    }
}

File: mpp/lab7/server/Repository/ParticipantRepository.cs
using System.Data.Common;
using lab5.utils;
using log4net;
using server.Domain;

namespace server.Repository
{
    public class ParticipantRepository : IParticipantRepository
    {
        private static readonly ILog Logger = LogManager.GetLogger("ParticipantRepository");
        
        private DbConnection Connection { get; }

        public ParticipantRepository(DbConnection connection)
        {
            Connection = connection;
            CreateTable();
        }

        private void CreateTable()
        {
            var command = Connection.CreateCommand(
                @"
                    create table if not exists Participants (
                        ParticipantId integer primary key autoincrement,
                        ParticipantName nvarchar(255) not null unique
                    );
                ");
            try
            {
                command.ExecuteNonQuery();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to create table");
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to create table", e);
            }
            
            Logger.Info("Created table or already existing");
            Logger.Info(command.CommandText);
        }

        public void Add(Participant participant)
        {
            var command = Connection.CreateCommand(
                @"
                    insert into Participants (
                        ParticipantName
                    ) values (
                        @ParticipantName
                    );
                ");
            command.AddParameterWithValue("@ParticipantName", participant.Name);
            try
            {
                command.ExecuteNonQuery();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to add participant");
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to add participant", e);
            }

            Logger.Info("Added participant");
            Logger.Info(command.CommandText);
            
            command = Connection.CreateCommand(
                @"
                    select
                        max(ParticipantId)
                    from Participants where
                        ParticipantName=@ParticipantName
                    ;
                ");

            command.AddParameterWithValue("@ParticipantName", participant.Name);

            DbDataReader reader;
            try
            {
                reader = command.ExecuteReader();
            }
            catch (DbException e)
            {
                Logger.Error("Failed to retrieve participant id");
                Logger.Error(command.CommandText);
                Logger.Error(e);
                throw new RepositoryError("Failed to retrieve participant id", e);
            }

            if (!reader.Read())
            {
                Logger.Error("Failed to retrieve participant id");
                Logger.Error(command.CommandText);
                throw new RepositoryError("Failed to retrieve participant id");
            }
            
            participant.Id = reader.GetInt32(0);

            Logger.Info("Retrieved participant id");
            Logger.Info(command.CommandText);
        }
    }
}
